plugins/askrene/algorithm.c

```diff
@@ -77,3 +77,95 @@ bool BFS_path(const tal_t *ctx, const struct graph *graph,
 	tal_free(this_ctx);
 	return target_found;
 }
+
+bool dijkstra_path(const tal_t *ctx, const struct graph *graph,
+		   const struct node source, const struct node destination,
+		   bool prune, const s64 *capacity, const s64 cap_threshold,
+		   const s64 *cost, const s64 *potential, struct arc *prev,
+		   s64 *distance)
+{
+	bool target_found = false;
+	const size_t max_num_arcs = graph_max_num_arcs(graph);
+	const size_t max_num_nodes = graph_max_num_nodes(graph);
+	tal_t *this_ctx = tal(ctx, tal_t);
+
+	/* check preconditions */
+	if (!graph || source.idx >=max_num_nodes || !cost || !capacity ||
+	    !prev || !distance)
+		goto finish;
+
```

Comment:
Assert, don't fail gracefully here.  Make sure caller never does this.  Once it's in the API as "permitted" it's really hard to remove.

Reply:
d24e60a643e377a91163afb0271d42958b04d027